### PR TITLE
Add lint and circleci, fix lint errors

### PR DESCRIPTION
**Pull request by baodongli**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+version: 2
+
+# Docker environment.
+defaults: &defaults
+  working_directory: /go/src/istio.io/istio
```
Comment (tiswanso) [inline .circleci/config.yml]: shouldn't this be `/go/src/istio.io/cni`?

**Comment by tiswanso** [comment]
+1 on @john-a-joyce 's comment.  Could you explain a bit more in the PR description?

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      - run: make lint
+
+workflows:
+  version: 2
+
+  all:
+    jobs:
+      - lint
+      - build
+
```
Comment (venilnoronha) [inline .circleci/config.yml]: nit: extra line feed

File: Makefile [hunk 3]
```diff
@@ -189,6 +189,27 @@ depend: init | $(ISTIO_OUT)
...
+#-----------------------------------------------------------------------------
+# Target: precommit
+#-----------------------------------------------------------------------------
+.PHONY: precommit format format.gofmt format.goimports lint buildcache
+
+# Target run by the pre-commit script, to automate formatting and lint
+# If pre-commit script is not used, please run this manually.
+precommit: format lint
+
+format:
```
Comment (venilnoronha) [inline Makefile]: I'd add a target description.

File: bin/check_license.sh [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  if head -20 "$fn" | grep "auto\\-generated" > /dev/null; then
+          continue
+  fi
+
+  if head -20 "$fn" | grep "DO NOT EDIT" > /dev/null; then
+          continue
+  fi
+
+  if head -20 "$fn" | grep "Code generated by go-bindata" > /dev/null; then
+          continue
```
Comment (venilnoronha) [inline bin/check_license.sh]: nit: inconsistent indentations here and above

File: bin/init_helm.sh [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Init script downloads or updates envoy and the go dependencies. Called from Makefile, which sets
+# the needed environment variables.
+
+
```
Comment (venilnoronha) [inline bin/init_helm.sh]: nit: extra line feed

File: bin/linters.sh [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+function run_gometalinter() {
+    echo 'Running gometalinter ....'
+    $gometalinter --config=./lintconfig_base.json ./...
+    echo 'gometalinter OK'
+}
+
+function run_helm_lint() {
+    echo 'Running helm lint on istio-cni'
+    helm lint ./deployments/kubernetes/install/helm/istio-cni
+    echo 'helm lint on istio-cni OK'
```
Comment (venilnoronha) [inline bin/linters.sh]: nit: this files has 3 spaced indentations, while rest have 2 space indentations

File: bin/init_helm.sh [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+if [ ! -f "${ISTIO_OUT}/version.helm.${HELM_VER}" ] ; then
+    TD=$(mktemp -d)
+    # Install helm. Please keep it in sync with .circleci
+    cd "${TD}" && \
+        curl -Lo "${TD}/helm.tgz" "https://storage.googleapis.com/kubernetes-helm/helm-${HELM_VER}-${LOCAL_OS}-amd64.tar.gz" && \
+        tar xfz helm.tgz && \
+        mv ${LOCAL_OS}-amd64/helm "${ISTIO_OUT}/helm-${HELM_VER}" && \
+        cp "${ISTIO_OUT}/helm-${HELM_VER}" "${ISTIO_OUT}/helm" && \
+        rm -rf "${TD}" && \
+        touch "${ISTIO_OUT}/version.helm.${HELM_VER}"
```
Comment (venilnoronha) [inline bin/init_helm.sh]: nit: inconsistent indentations

File: bin/linters.sh (hunk 6 above)
Comment (tiswanso) [reply bin/linters.sh]: @venilnoronha  -- I'm betting @baodongli copied this file and the other .sh files from istio/istio, in which case I don't think changing the spacing in these is preferable.

Comment (venilnoronha) [reply bin/linters.sh]: @tiswanso got you.

@baodongli please ignore my comments on this file.

File: bin/init_helm.sh (hunk 7 above)
Comment (venilnoronha) [reply bin/init_helm.sh]: Ignore

File: Makefile (hunk 3 above)
Comment (venilnoronha) [reply Makefile]: Ignore.

File: bin/check_license.sh (hunk 4 above)
Comment (venilnoronha) [reply bin/check_license.sh]: Ignore.

File: bin/init_helm.sh (hunk 5 above)
Comment (venilnoronha) [reply bin/init_helm.sh]: Ignore.

File: bin/linters.sh (hunk 6 above)
Comment (baodongli) [reply bin/linters.sh]: Yeah, these files are copied and modified for istio-cni. When istio-cni is moved under istio, these files will most likely be thrown away, and the original files be updated to support istio-cni

**Pull request merged**